### PR TITLE
NOISSUE: Refactor

### PR DESCRIPTION
**Pull request by mteodor**

### CODE REVIEW

**Comment by drasko** [comment]
Please resolve the conflicts and update

File: cmd/main.go [hunk 1]
```diff
@@ -80,7 +80,8 @@ const (
 )
 
 func main() {
-	logger, err := logger.New(os.Stdout, defLogLevel)
+	logLvl := mainflux.Env(envLogLevel, defLogLevel)
```
Comment (manuio) [inline cmd/main.go:83]: I would be better if you can do first `loadConfig()` and then use `cfg.logLevel`. For that you could create a function to load agent conf and where you need the logger.

File: pkg/agent/service.go [hunk 2]
```diff
@@ -137,22 +171,22 @@ func New(mc paho.Client, cfg *config.Config, ec edgex.Client, nc *nats.Conn, log
 
 func (a *agent) Execute(uuid, cmd string) (string, error) {
 	cmdArr := strings.Split(strings.Replace(cmd, " ", "", -1), ",")
-	if len(cmd) < 1 {
+	if len(cmdArr) < 1 {
```
Comment (manuio) [inline pkg/agent/service.go:174]: If you use cmdArr[1:] probably here `len(cmdArr) < 2`

File: pkg/config/config.go [hunk 3]
```diff
@@ -89,7 +89,7 @@ func (c *Config) Save() errors.Error {
 }
 
 // Read - retrieve config from a file
-func (c *Config) Read() errors.Error {
+func (c *Config) Read() error {
```
Comment (manuio) [inline pkg/config/config.go:92]: I'm not sure but maybe here you can return Config instead of use a pointer

File: pkg/agent/service.go [hunk 4]
```diff
@@ -223,14 +257,85 @@ func (a *agent) ServiceConfig(uuid, cmdStr string) error {
...
+		}
+	}
+	return nil
+}
+
+func (a *agent) terminalOpen(uuid string) error {
+	if _, ok := a.terminals[uuid]; !ok {
+		term, err := terminal.NewSession(uuid, a.Publish, a.logger)
+		if err != nil {
+			return errors.Wrap(errors.Wrap(errFailedToCreateTerminalSession, fmt.Errorf("failed for %s", uuid)), err)
```
Comment (manuio) [inline pkg/agent/service.go:296]: This second `failed for` is redundant. Maybe `for uuid %s`?

File: pkg/config/config.go (hunk 3 above)
Comment (mteodor) [reply pkg/config/config.go:92]: large change, out of scope

File: cmd/main.go (hunk 1 above)
Comment (mteodor) [reply cmd/main.go:83]: I need logger before I can get configuration from the bootstrap service when I'm actually getting the conf file for agent , so I'll create logger with env and then after succsefull bootstrap ill recreate logger

Comment (manuio) [reply cmd/main.go:83]: What I meant is that you should have in the main something like:
```
cfg := config {
	logLevel:   mainflux.Env(envLogLevel, defLogLevel),
	natsURL:   mainflux.Env(envNatsURL, defNatsURL),
	httpPort:   mainflux.Env(envHTTPPort, defHTTPPort),
	configFile: mainflux.Env(envConfigFile, defConfigFile),
}
```
This should be set with loadConfig().

You should then instantiate the logger and pass `cfg` and `logger` to `loadBootstrapConfig(cfg, logger)`

File: cmd/main.go [hunk 5]
```diff
@@ -79,17 +80,26 @@ const (
...
-	logger, err := logger.New(os.Stdout, logLvl)
+	cfg, err := loadEnvConfig()
+	if err != nil {
+		log.Fatalf(fmt.Sprintf("Failed to load config: %s", err.Error()))
+	}
+
+	logger, err := logger.New(os.Stdout, cfg.Agent.Log.Level)
 	if err != nil {
-		log.Fatalf(fmt.Sprintf("Failed to create logger: %s", err.Error()))
+		log.Fatalf(fmt.Sprintf("Failed create logger: %s", err.Error()))
```
Comment (manuio) [inline cmd/main.go:97]: `Failed to create logger`

File: pkg/bootstrap/bootstrap.go [hunk 6]
```diff
@@ -178,7 +177,8 @@ func getConfig(bsID, bsKey, bsSvrURL string, logger log.Logger) (deviceConfig, e
 		return deviceConfig{}, errors.New(err.Error())
 	}
 	defer resp.Body.Close()
-
+	fmt.Printf("url:%s", url)
+	fmt.Printf("body:%s", string(body))
```
Comment (manuio) [inline pkg/bootstrap/bootstrap.go:181]: rm this logs

**Pull request merged**